Podcast Planner verbiage and layout

brandonhundt opened the pull request:
closes: #606 

Now that we have a working calendar (!!!EXCITING!!!) I thought we could take a pass at the verbiage and the layout of the calendar. This PR's changes will get us closer to what exists in Publish currently where we've received a lot of neutral to positive feedback. (NOTE: I think we can open a new ticket to explore enhancements to this feature that go beyond the existing functionality in this PR.)

There were a lot of places where we had help text, headers, and field labels that felt a little redundant or overly technical in its use, so I attempted to clean It up.

Also, I reverted this to a single column layout like we have now. I think this will serve our producers finely enough until we get the time to really rethink the feature in the future.

brandonhundt commented on app/views/podcast_planner/_form_date_calculator.html.erb:
```diff
@@ -1,12 +1,6 @@
-  <div class="mb-4">
-    <h3 class="fw-bold"><%= t(".header.day_select") %></h3>
-    <p><%= t(".help.day_select") %></p>
+    <h3 class="mb-3"><%= t(".header.day_select") %></h3>
```
Example of simplifying the text presented to the user.

brandonhundt commented on app/views/podcast_planner/_form_date_calculator.html.erb:
```diff
@@ -46,25 +36,26 @@
+          <%= form.date_field :start_date, min: DateTime.tomorrow, max: DateTime.tomorrow + 1.years, value: DateTime.tomorrow, data: {action: "click#submit"} %>
+          <%= form.label :start_date %>
+        </div>
+      </div>
```
I adjusted the layout of this field so that the selector field is full width, with two variable fields appearing below. This is more consistent with our UI patterns.

brandonhundt commented on app/views/podcast_planner/_form_main.html.erb:
```diff
@@ -8,7 +8,7 @@
     ) do |form| %>
     <h1 class="text-black fw-bold d-flex flex-fill"><%= t(".header") %></h1>
-    <p><%= t(".help_html", url: new_podcast_episode_path(@podcast)) %></p>
+    <p class="lh-sm"><%= t(".help_html", url: new_podcast_episode_path(@podcast)) %></p>
```
Line-height was really tall, looked weird.

brandonhundt commented on app/views/podcast_planner/_form_side.html.erb:
```diff
@@ -1,56 +1,51 @@
+        <%= form.select :publish_time, time_options %>
+        <%= form.label :publish_time %>
+      </div>
+    </div>
```
This file was reworked since its no longer technically a sidebar card

brandonhundt replied:
Functionally there shouldn't be any different. All presentational here.

brandonhundt commented on config/locales/en.yml:
```diff
@@ -230,13 +230,13 @@ en:
     form_main:
       header: Plan Episodes
-      help_html: Save yourself some time by automatically creating multiple episode drafts to create a production schedule, it may also help your revenue as our sales teams will have access to dropdates to sell against. If you only wish to add a single episode, <a href="%{url}">go here</a>.
+      help_html: Add multiple episode drafts to create a production schedule. Episode drafts improve the accuracy of your podcast's forecasts. If you only need to add a single episode, <a href="%{url}">go here</a>.
```
A verbiage review in this file is very appreciated!!!

radical-ube commented on app/views/podcast_planner/_form_side.html.erb:
```diff
@@ -1,56 +1,51 @@
+  <div class="card p-4 mt-4mb-4">
+    <div data-controller="pager" data-pager-range-value="6">
+      <div class="d-flex justify-content-between">
+        <h3 class="mb-3"><%= t(".header.number_of_drafts") %></h3>
```
this needs a `count` variable to pass through to the yml file in order to display the correct number.

you can probably pass through `@planner.dates.count` but that would only reflect the dates generated through the calculator, not selected manually. i could probably figure out a helper method to add to the stimulus controller that would update every time a manual date has been clicked as well. is this an important feature for parity?

radical-ube reviewed (commented):
in general i have no comments on verbiage since this is a return to parity verbiage and we're not planning on the improvements until post-launch. the only comment i made was regarding the `count` variable for that one header label. the only other comment i would make is figuring out a color key since that is one thing new to the calendar.

radical-ube reviewed (changes requested):
should have put previous comment under request changes for the `count` variable.

radical-ube commented on app/helpers/podcast_planner_helper.rb:
```diff
@@ -5,6 +5,7 @@ module PodcastPlannerHelper
   MONTHLY_WEEKS = I18n.t([:first, :second, :third, :fourth, :fifth], scope: [:podcast_planner, :helper, :monthly_options])
   DATE_CONTROLLER = "date"
   TOGGLE_ACTION = "click->date#toggleSelect"
+  RECOUNT_ACTION = "click->count#recount"
```
this and the `counterTarget` below are additions for the new stimulus `count_controller`. Essentially adds a counter to each date. the controller counts the ones that are selected.

radical-ube replied:
moved `counterTarget` to the input in the `_calendar` partial.

radical-ube commented on app/javascript/controllers/count_controller.js:
```diff
@@ -0,0 +1,25 @@
+    this.displayTarget.innerHTML = this.selectedValue
+  }
+
+  countSelectedTargets() {
```
i wasn't quite sure how to generalize this action, so for now it is reading the `disabled` property on the inputs within the dates. and then calculating which dates are selected by determining which have been enabled.

cavis replied:
It is a little awkward to not just count the `counterTargets`.

Could you instead add this to the `form.hidden_field` in `_calendar.html.erb`?  Would basically look the same, except getting rid of the `firstElementChild`.  Maybe a bit cleaner?

Or it's nearly the same - your call.

radical-ube commented on config/locales/en.yml:
```diff
@@ -248,16 +248,19 @@ en:
+        publish_time: What time of day do you plan to publish the episodes?
+        audio_segments: How many audio segments?
+        segment_count: Number of segments
+        number_of_drafts: "Drafts planned: "
```
changed this so it no longer needs to pass in a `count` variable. the count is displayed in the view in the innerHTML of a new span in the h3 that uses this label.

radical-ube commented on app/views/podcast_planner/_form_side.html.erb:
```diff
@@ -1,56 +1,53 @@
+      <div data-controller="pager count" data-pager-range-value="6">
+        <div class="d-flex justify-content-between">
+          <h3 class="mb-3"><%= t(".header.number_of_drafts") %>
+            <span data-count-target="display"></span>
```
new span here. the innerHTML of this will update with the number of selected items, as determined by the new `count_controller`.

cavis replied:
I would probably render a number in here to start?  So it has the correct value in HTML, before javascript even takes a pass at it.

cavis commented on app/javascript/controllers/count_controller.js:
```diff
@@ -0,0 +1,25 @@
+    const selectedTargets = this.counterTargets.filter((el) => {
+      return el.firstElementChild.disabled
+    })
+    return this.counterTargets.length - selectedTargets.length
```
Naming is a bit misleading ... `selectedTargets` appears to be all the _deselected_ hidden fields?  Maybe invert the above filter to `!disabled`, and then just return `.length`?

cavis commented on app/javascript/controllers/count_controller.js:
```diff
@@ -0,0 +1,25 @@
+export default class extends Controller {
+  static targets = ["counter", "display"]
+  static values = {
+    selected: { type: Number, default: 0 },
```
Is this even being passed in anywhere, or are you getting the default?

I'd probably suggest omitting this entirely.  The server should render the correct number in the HTML to start with.  And then drop the `connect()` method here, and just wait until some action calls `recount()` to have this controller modify the innerHTML.

cavis reviewed (changes requested):
Couple minor suggestions @radical-ube.  But all pretty much your call - they're slightly stylistic.

@brandonhundt one slight issue on wide viewports 
![image](https://user-images.githubusercontent.com/1410587/235775786-78b55e02-2c7c-437c-8dfe-12a7e15aa1f6.png)

It isn't going to be trivial to display 8 calendars magically as the user widens the viewport.  So may just want to restrict the max-width of those for now?  Or force maximum 3-per-row?

brandonhundt commented:
@ube @ryan I've fixed the grid issue so that on the largest screens it will always display in 3 columns.

Also added a key for the calendar, and tweaked the hover state on the calendar dates.

![Screenshot 2023-05-03 at 1 02 18 PM](https://user-images.githubusercontent.com/369739/236005184-f37f8581-22f9-4eed-b51f-bf908ddaab6c.png)

brandonhundt commented on app/assets/stylesheets/app/calendar.scss:
```diff
@@ -19,8 +19,9 @@
-      background-color: $secondary !important;
+      background-color: $primary !important;
       color: $light;
+      cursor: pointer;
```
to let the user know they can click to add dates. chose the link color to emphasize clicking.

brandonhundt commented on app/views/podcast_planner/_calendar.html.erb:
```diff
@@ -1,9 +1,9 @@
-<div class="col <%= "d-none" if index >= range %>" data-pager-target="page">
+<div class="col col-xl-4 <%= "d-none" if index >= range %>" data-pager-target="page">
```
Thats all it was to fix the grid issue.

cavis reviewed (approved):
👍
